pupil_src/shared_modules/annotations.py

```diff
@@ -7,239 +7,329 @@
+        for name, hotkey in self._initial_annotation_definitions:
+            self._add_annotation_definition(name, hotkey)
+
+    @abstractmethod
+    def customize_menu(self):
+        pass
 
     def deinit_ui(self):
-        self.menu.remove(self.sub_menu)
-        self.sub_menu = None
+        self._clear_button_quickbars()
         self.remove_menu()
-        if self.buttons:
-            for b in self.buttons:
-                self.g_pool.quickbar.remove(b)
-            self.buttons = []
-
-    def append_annotation(self):
-        self.annotation_definitions.append((self.new_annotation_name,
-                                            self.new_annotation_hotkey))
-        self.update_buttons()
-
-    def remove_annotation(self, annotation):
-        try:
-            self.annotation_definitions.remove(annotation)
-        except ValueError:
-            print(annotation, self.annotation_definitions)
-        self.update_buttons()
 
-    def fire_annotation(self, annotation_label):
-        t = self.g_pool.get_timestamp()
-        logger.info('"{}"@{}'.format(annotation_label, t))
-        # you may add more field to this dictionary if you want.
-        notification = {'subject': 'annotation', 'label': annotation_label,
-                        'timestamp': t, 'duration': 0.0, 'record': True}
-        self.notify_all(notification)
+    def _clear_button_quickbars(self):
+        for annotation_info in self._annotation_infos:
+            self._remove_button_quickbar(annotation_info)
+
+    @abstractmethod
+    def fire_annotation(self, annotation_definition):
+        pass
+
+    def _add_annotation_clicked(self):
+        # new_annotation_name and hotkey are set by the ui input fields
+        self._add_annotation_definition(
+            self._new_annotation_name, self._new_annotation_hotkey
+        )
+
+    def _add_annotation_definition(self, annotation_name, hotkey):
+        annotation_definition = AnnotationDefinition(
+            name=annotation_name, hotkey=hotkey
+        )
+        annotation_info = self._AnnotationInfo(annotation_definition)
+        annotation_info.button_quickbar = self._create_button_quickbar(
+            annotation_definition
+        )
+        annotation_info.button_menu = self._create_button_menu(annotation_info)
+        self._annotation_infos.append(annotation_info)
+
+        self._append_button_quickbar(annotation_info)
+        self._append_button_menu(annotation_info)
+
+    def _create_button_quickbar(self, annotation_definition):
+        def make_fire(definition):
+            return lambda _: self.fire_annotation(definition)
+
+        return ui.Thumb(
+            annotation_definition.name,
+            setter=make_fire(annotation_definition),
+            getter=lambda: False,
+            label=annotation_definition.hotkey,
+            hotkey=annotation_definition.hotkey,
+        )
+
+    def _create_button_menu(self, annotation_info):
+        def make_remove(info):
```

Comment:
wouldn't `make_remove = lambda: self._remove_annotation_info(annotation_info)` be simpler? And later 
`return ui.Button(label=name + " <" + hotkey + ">", function=make_remove)`

Reply:
This would not be the same! This is a so called [closure](https://www.programiz.com/python-programming/closure).